// @3-/coffee_plus/index.cjs
#!/usr/bin/env -S node --loader=@w5/jsext --trace-uncaught --expose-gc --unhandled-rejections=strict --experimental-import-meta-resolve
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// node_modules/@3-/camel/index.js
var camel_default;
var init_camel = __esm({
  "node_modules/@3-/camel/index.js"() {
    camel_default = (s) => {
      var i, j, len, pre_is_up, t, up, up_case;
      if (s.length) {
        t = [];
        for (j = 0, len = s.length; j < len; j++) {
          i = s[j];
          if (i === "-") {
            up = true;
            continue;
          }
          if (i === "_") {
            if (!pre_is_up) {
              up = true;
              continue;
            }
          }
          up_case = i.toUpperCase();
          if (up) {
            i = up_case;
            up = void 0;
            pre_is_up = true;
          } else if ("9876543210".includes(i)) {
            pre_is_up = false;
          } else {
            pre_is_up = i === up_case;
          }
          t.push(i);
        }
        return t.join("");
      } else {
        return s;
      }
    };
  }
});

// node_modules/lodash-es/_freeGlobal.js
var freeGlobal, freeGlobal_default;
var init_freeGlobal = __esm({
  "node_modules/lodash-es/_freeGlobal.js"() {
    freeGlobal = typeof global == "object" && global && global.Object === Object && global;
    freeGlobal_default = freeGlobal;
  }
});

// node_modules/lodash-es/_root.js
var freeSelf, root, root_default;
var init_root = __esm({
  "node_modules/lodash-es/_root.js"() {
    init_freeGlobal();
    freeSelf = typeof self == "object" && self && self.Object === Object && self;
    root = freeGlobal_default || freeSelf || Function("return this")();
    root_default = root;
  }
});

// node_modules/lodash-es/_Symbol.js
var Symbol2, Symbol_default;
var init_Symbol = __esm({
  "node_modules/lodash-es/_Symbol.js"() {
    init_root();
    Symbol2 = root_default.Symbol;
    Symbol_default = Symbol2;
  }
});

// node_modules/lodash-es/_getRawTag.js
function getRawTag(value) {
  var isOwn = hasOwnProperty.call(value, symToStringTag), tag = value[symToStringTag];
  try {
    value[symToStringTag] = void 0;
    var unmasked = true;
  } catch (e) {
  }
  var result = nativeObjectToString.call(value);
  if (unmasked) {
    if (isOwn) {
      value[symToStringTag] = tag;
    } else {
      delete value[symToStringTag];
    }
  }
  return result;
}
var objectProto, hasOwnProperty, nativeObjectToString, symToStringTag, getRawTag_default;
var init_getRawTag = __esm({
  "node_modules/lodash-es/_getRawTag.js"() {
    init_Symbol();
    objectProto = Object.prototype;
    hasOwnProperty = objectProto.hasOwnProperty;
    nativeObjectToString = objectProto.toString;
    symToStringTag = Symbol_default ? Symbol_default.toStringTag : void 0;
    getRawTag_default = getRawTag;
  }
});

// node_modules/lodash-es/_objectToString.js
function objectToString(value) {
  return nativeObjectToString2.call(value);
}
var objectProto2, nativeObjectToString2, objectToString_default;
var init_objectToString = __esm({
  "node_modules/lodash-es/_objectToString.js"() {
    objectProto2 = Object.prototype;
    nativeObjectToString2 = objectProto2.toString;
    objectToString_default = objectToString;
  }
});

// node_modules/lodash-es/_baseGetTag.js
function baseGetTag(value) {
  if (value == null) {
    return value === void 0 ? undefinedTag : nullTag;
  }
  return symToStringTag2 && symToStringTag2 in Object(value) ? getRawTag_default(value) : objectToString_default(value);
}
var nullTag, undefinedTag, symToStringTag2, baseGetTag_default;
var init_baseGetTag = __esm({
  "node_modules/lodash-es/_baseGetTag.js"() {
    init_Symbol();
    init_getRawTag();
    init_objectToString();
    nullTag = "[object Null]";
    undefinedTag = "[object Undefined]";
    symToStringTag2 = Symbol_default ? Symbol_default.toStringTag : void 0;
    baseGetTag_default = baseGetTag;
  }
});

// node_modules/lodash-es/isObjectLike.js
function isObjectLike(value) {
  return value != null && typeof value == "object";
}
var isObjectLike_default;
var init_isObjectLike = __esm({
  "node_modules/lodash-es/isObjectLike.js"() {
    isObjectLike_default = isObjectLike;
  }
});

// node_modules/lodash-es/isSymbol.js
function isSymbol(value) {
  return typeof value == "symbol" || isObjectLike_default(value) && baseGetTag_default(value) == symbolTag;
}
var symbolTag, isSymbol_default;
var init_isSymbol = __esm({
  "node_modules/lodash-es/isSymbol.js"() {
    init_baseGetTag();
    init_isObjectLike();
    symbolTag = "[object Symbol]";
    isSymbol_default = isSymbol;
  }
});

// node_modules/lodash-es/_arrayMap.js
function arrayMap(array, iteratee) {
  var index = -1, length = array == null ? 0 : array.length, result = Array(length);
  while (++index < length) {
    result[index] = iteratee(array[index], index, array);
  }
  return result;
}
var arrayMap_default;
var init_arrayMap = __esm({
  "node_modules/lodash-es/_arrayMap.js"() {
    arrayMap_default = arrayMap;
  }
});

// node_modules/lodash-es/isArray.js
var isArray, isArray_default;
var init_isArray = __esm({
  "node_modules/lodash-es/isArray.js"() {
    isArray = Array.isArray;
    isArray_default = isArray;
  }
});

// node_modules/lodash-es/_baseToString.js
function baseToString(value) {
  if (typeof value == "string") {
    return value;
  }
  if (isArray_default(value)) {
    return arrayMap_default(value, baseToString) + "";
  }
  if (isSymbol_default(value)) {
    return symbolToString ? symbolToString.call(value) : "";
  }
  var result = value + "";
  return result == "0" && 1 / value == -INFINITY ? "-0" : result;
}
var INFINITY, symbolProto, symbolToString, baseToString_default;
var init_baseToString = __esm({
  "node_modules/lodash-es/_baseToString.js"() {
    init_Symbol();
    init_arrayMap();
    init_isArray();
    init_isSymbol();
    INFINITY = 1 / 0;
    symbolProto = Symbol_default ? Symbol_default.prototype : void 0;
    symbolToString = symbolProto ? symbolProto.toString : void 0;
    baseToString_default = baseToString;
  }
});

// node_modules/lodash-es/toString.js
function toString(value) {
  return value == null ? "" : baseToString_default(value);
}
var toString_default;
var init_toString = __esm({
  "node_modules/lodash-es/toString.js"() {
    init_baseToString();
    toString_default = toString;
  }
});

// node_modules/lodash-es/_baseSlice.js
function baseSlice(array, start, end) {
  var index = -1, length = array.length;
  if (start < 0) {
    start = -start > length ? 0 : length + start;
  }
  end = end > length ? length : end;
  if (end < 0) {
    end += length;
  }
  length = start > end ? 0 : end - start >>> 0;
  start >>>= 0;
  var result = Array(length);
  while (++index < length) {
    result[index] = array[index + start];
  }
  return result;
}
var baseSlice_default;
var init_baseSlice = __esm({
  "node_modules/lodash-es/_baseSlice.js"() {
    baseSlice_default = baseSlice;
  }
});

// node_modules/lodash-es/_castSlice.js
function castSlice(array, start, end) {
  var length = array.length;
  end = end === void 0 ? length : end;
  return !start && end >= length ? array : baseSlice_default(array, start, end);
}
var castSlice_default;
var init_castSlice = __esm({
  "node_modules/lodash-es/_castSlice.js"() {
    init_baseSlice();
    castSlice_default = castSlice;
  }
});

// node_modules/lodash-es/_hasUnicode.js
function hasUnicode(string) {
  return reHasUnicode.test(string);
}
var rsAstralRange, rsComboMarksRange, reComboHalfMarksRange, rsComboSymbolsRange, rsComboRange, rsVarRange, rsZWJ, reHasUnicode, hasUnicode_default;
var init_hasUnicode = __esm({
  "node_modules/lodash-es/_hasUnicode.js"() {
    rsAstralRange = "\\ud800-\\udfff";
    rsComboMarksRange = "\\u0300-\\u036f";
    reComboHalfMarksRange = "\\ufe20-\\ufe2f";
    rsComboSymbolsRange = "\\u20d0-\\u20ff";
    rsComboRange = rsComboMarksRange + reComboHalfMarksRange + rsComboSymbolsRange;
    rsVarRange = "\\ufe0e\\ufe0f";
    rsZWJ = "\\u200d";
    reHasUnicode = RegExp("[" + rsZWJ + rsAstralRange + rsComboRange + rsVarRange + "]");
    hasUnicode_default = hasUnicode;
  }
});

// node_modules/lodash-es/_asciiToArray.js
function asciiToArray(string) {
  return string.split("");
}
var asciiToArray_default;
var init_asciiToArray = __esm({
  "node_modules/lodash-es/_asciiToArray.js"() {
    asciiToArray_default = asciiToArray;
  }
});

// node_modules/lodash-es/_unicodeToArray.js
function unicodeToArray(string) {
  return string.match(reUnicode) || [];
}
var rsAstralRange2, rsComboMarksRange2, reComboHalfMarksRange2, rsComboSymbolsRange2, rsComboRange2, rsVarRange2, rsAstral, rsCombo, rsFitz, rsModifier, rsNonAstral, rsRegional, rsSurrPair, rsZWJ2, reOptMod, rsOptVar, rsOptJoin, rsSeq, rsSymbol, reUnicode, unicodeToArray_default;
var init_unicodeToArray = __esm({
  "node_modules/lodash-es/_unicodeToArray.js"() {
    rsAstralRange2 = "\\ud800-\\udfff";
    rsComboMarksRange2 = "\\u0300-\\u036f";
    reComboHalfMarksRange2 = "\\ufe20-\\ufe2f";
    rsComboSymbolsRange2 = "\\u20d0-\\u20ff";
    rsComboRange2 = rsComboMarksRange2 + reComboHalfMarksRange2 + rsComboSymbolsRange2;
    rsVarRange2 = "\\ufe0e\\ufe0f";
    rsAstral = "[" + rsAstralRange2 + "]";
    rsCombo = "[" + rsComboRange2 + "]";
    rsFitz = "\\ud83c[\\udffb-\\udfff]";
    rsModifier = "(?:" + rsCombo + "|" + rsFitz + ")";
    rsNonAstral = "[^" + rsAstralRange2 + "]";
    rsRegional = "(?:\\ud83c[\\udde6-\\uddff]){2}";
    rsSurrPair = "[\\ud800-\\udbff][\\udc00-\\udfff]";
    rsZWJ2 = "\\u200d";
    reOptMod = rsModifier + "?";
    rsOptVar = "[" + rsVarRange2 + "]?";
    rsOptJoin = "(?:" + rsZWJ2 + "(?:" + [rsNonAstral, rsRegional, rsSurrPair].join("|") + ")" + rsOptVar + reOptMod + ")*";
    rsSeq = rsOptVar + reOptMod + rsOptJoin;
    rsSymbol = "(?:" + [rsNonAstral + rsCombo + "?", rsCombo, rsRegional, rsSurrPair, rsAstral].join("|") + ")";
    reUnicode = RegExp(rsFitz + "(?=" + rsFitz + ")|" + rsSymbol + rsSeq, "g");
    unicodeToArray_default = unicodeToArray;
  }
});

// node_modules/lodash-es/_stringToArray.js
function stringToArray(string) {
  return hasUnicode_default(string) ? unicodeToArray_default(string) : asciiToArray_default(string);
}
var stringToArray_default;
var init_stringToArray = __esm({
  "node_modules/lodash-es/_stringToArray.js"() {
    init_asciiToArray();
    init_hasUnicode();
    init_unicodeToArray();
    stringToArray_default = stringToArray;
  }
});

// node_modules/lodash-es/_createCaseFirst.js
function createCaseFirst(methodName) {
  return function(string) {
    string = toString_default(string);
    var strSymbols = hasUnicode_default(string) ? stringToArray_default(string) : void 0;
    var chr = strSymbols ? strSymbols[0] : string.charAt(0);
    var trailing = strSymbols ? castSlice_default(strSymbols, 1).join("") : string.slice(1);
    return chr[methodName]() + trailing;
  };
}
var createCaseFirst_default;
var init_createCaseFirst = __esm({
  "node_modules/lodash-es/_createCaseFirst.js"() {
    init_castSlice();
    init_hasUnicode();
    init_stringToArray();
    init_toString();
    createCaseFirst_default = createCaseFirst;
  }
});

// node_modules/lodash-es/upperFirst.js
var upperFirst, upperFirst_default;
var init_upperFirst = __esm({
  "node_modules/lodash-es/upperFirst.js"() {
    init_createCaseFirst();
    upperFirst = createCaseFirst_default("toUpperCase");
    upperFirst_default = upperFirst;
  }
});

// node_modules/lodash-es/lodash.js
var init_lodash = __esm({
  "node_modules/lodash-es/lodash.js"() {
    init_upperFirst();
  }
});

// lib/import.js
var import_exports = {};
__export(import_exports, {
  default: () => import_default
});
function* import_default(code) {
  var bk, i, in_gt, trim;
  in_gt = false;
  bk = [];
  for (i of code) {
    trim = i.trimEnd();
    if (trim) {
      if (i.startsWith(">")) {
        in_gt = true;
        bk.push(i);
      } else if (in_gt) {
        if (i.charAt(0) !== " ") {
          in_gt = false;
          yield* _import(bk);
          bk = [];
          yield i;
        } else {
          bk.push(i);
        }
      } else {
        yield i;
      }
    } else {
      yield i;
    }
  }
  if (bk.length) {
    return yield* _import(bk);
  }
}
var BUILDIN, CamelCase, _buildin, _import, _import_line, _mod_name, hash;
var init_import = __esm({
  "lib/import.js"() {
    init_camel();
    init_lodash();
    CamelCase = (x) => {
      var r;
      r = camel_default(x);
      if (r !== x) {
        return upperFirst_default(r);
      } else {
        return x;
      }
    };
    _mod_name = (i, convert = CamelCase) => {
      var dot, ext, name, pos;
      pos = i.lastIndexOf(":");
      if (pos < 0) {
        name = i.split("/").pop();
        dot = name.lastIndexOf(".");
        if (dot > 0) {
          ext = name.slice(dot + 1);
          if (~["css", "styl"].indexOf(ext)) {
            name = "";
          } else {
            name = name.slice(0, dot);
          }
        }
        return [convert(name), i];
      } else {
        return [i.slice(pos + 1), _buildin(i, pos)];
      }
    };
    _import_line = function(code) {
      var mod, name;
      [name, mod] = _mod_name(code);
      if (name) {
        return `import ${name} from '${mod}'`;
      } else {
        return `import '${mod}'`;
      }
    };
    hash = "#";
    BUILDIN = ["bun", "npm", "node"];
    _buildin = (mod, pos) => {
      var _mod;
      _mod = mod.slice(0, pos);
      if (BUILDIN.indexOf(_mod) < 0) {
        return _mod;
      } else {
        return mod;
      }
    };
    _import = function* (bk) {
      var func, func_li, gt_pos, i, import_li, j, len, li, mod, pos, r, ref, rename;
      for (li of bk) {
        if (li.charAt(0) === hash) {
          yield li;
          continue;
        }
        li = li.slice(1).trim();
        if (li) {
          if (li.startsWith(hash)) {
            yield li;
            continue;
          }
          pos = li.lastIndexOf(hash);
          if (pos > 0) {
            li = li.slice(0, pos).trimEnd();
          }
          li = li.split(" ");
          gt_pos = li.indexOf(">");
          if (gt_pos > 0) {
            import_li = li.slice(0, gt_pos - 1);
          } else {
            import_li = li;
          }
          r = import_li.map(_import_line);
          if (gt_pos > 0) {
            mod = li[gt_pos - 1];
            pos = mod.lastIndexOf(":");
            if (pos > 0) {
              rename = mod.slice(pos + 1);
              mod = mod.slice(0, pos);
              if (BUILDIN.indexOf(mod) < 0) {
                if (rename === "@") {
                  rename = mod.split("/").pop().split(".")[0];
                  rename = CamelCase(rename);
                }
                r.push(`import ${rename} from '${mod}'`);
              }
            }
            func_li = [];
            ref = li.slice(gt_pos + 1);
            for (j = 0, len = ref.length; j < len; j++) {
              i = ref[j];
              [rename, func] = _mod_name(i, (x) => {
                return x;
              });
              if (rename === func) {
                func_li.push(func);
              } else {
                func_li.push(`${func} as ${rename}`);
              }
            }
            if (func_li.length > 0) {
              r.push(`import {${func_li.join(",")}} from '${mod}'`);
            }
          }
          yield r.join(";");
        } else {
          yield "";
        }
      }
    };
  }
});

// lib/index.cjs
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.coffee_plus = void 0;
var _import2 = _interopRequireDefault((init_import(), __toCommonJS(import_exports)));
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}
var coffee_label;
var label;
coffee_label = (code) => {
  var _export, argp, c0, eqp, i, indent, j, len, li, line, pos, pre, pre_indent, ref, s, word;
  li = [];
  pre_indent = 0;
  pre = 0;
  OUT:
    for (line of code) {
      s = line.trimStart();
      indent = line.length - s.length;
      if (pre) {
        li.push("".padEnd(pre_indent) + "if null #" + pre.trim());
        pre = 0;
      }
      if (s.startsWith(":")) {
        pos = s.search(/\s/);
        if (~pos) {
          li.push("".padEnd(indent) + "$|" + s.slice(pos));
        } else {
          pre_indent = indent;
          pre = line;
        }
      } else {
        indent = "".padEnd(indent);
        c0 = s.charAt(0);
        if (s.startsWith("@->") || s.startsWith("@(") && s.endsWith(")->")) {
          li.push(indent + "constructor:" + s.slice(1));
        } else if (s.startsWith("< class ")) {
          li.push(indent + "export" + s.slice(1));
        } else if (~"+<".indexOf(c0) && " " === s.charAt(1)) {
          pos = s.indexOf("=>");
          if (pos < 0) {
            pos = s.indexOf("->");
            if (pos < 0) {
              pos = s.indexOf("{");
              if (pos < 0) {
                pos = s.indexOf("new ");
                if (pos < 0) {
                  pos = s.indexOf("[");
                }
              }
            }
          }
          if (pos > 0) {
            _export = "export ";
            eqp = s.indexOf("=");
            if (eqp < 0 || eqp === pos) {
              _export += "default ";
            } else {
              argp = s.indexOf("(");
              if (argp > 0 && eqp > argp) {
                _export += "default ";
              }
            }
            li.push(_export + s.slice(1).trimStart());
          } else {
            s = s.slice(2).split(",").map((i2) => {
              return i2.trim();
            }).filter(Boolean);
            for (j = 0, len = s.length; j < len; j++) {
              i = s[j];
              pos = i.indexOf("=");
              if (pos < 0) {
                i += "=`undefined`#" + c0;
              }
              if (c0 === "<") {
                i = "export " + i;
              }
              li.push(indent + i);
            }
          }
        } else {
          ref = ["break ", "continue "];
          for (word of ref) {
            if (s.startsWith(word)) {
              li.push(line.replace(word, "`" + word) + "`");
              continue OUT;
            }
          }
          li.push(line);
        }
      }
    }
  return li.join("\n");
};
label = (code) => {
  var indent, j, len, li, line, push, s, txt;
  li = [];
  code = code.split("\n");
  if (code[0] === "//!/usr/bin/env coffee") {
    li.push("#!/usr/bin/env -S node --import=@3-/jsext --trace-uncaught --expose-gc --unhandled-rejections=strict");
    code = code.slice(1);
  }
  for (j = 0, len = code.length; j < len; j++) {
    line = code[j];
    s = line.trimStart();
    indent = line.length - s.length;
    push = (txt2) => {
      li.push("".padEnd(indent) + txt2);
    };
    if (s.startsWith("export var ") && s.endsWith("=> {")) {
      li.push("export const" + s.slice(10));
    } else if (s.endsWith(" = undefined; //+")) {
      continue;
    } else if (s.endsWith(" = undefined; //<")) {
      li.push(line.slice(0, -17) + ";");
    } else if (s.startsWith("if (null) { //:")) {
      push(s.slice(15) + " : {");
    } else if (s.startsWith("$ | (")) {
      s = s.slice(4);
      if (s.indexOf("function(") > 0 || s.indexOf("=>") > 0) {
        txt = s;
      } else {
        txt = s.slice(1, -2) + ";";
      }
      push("$ : " + txt);
    } else if (s.startsWith("(() => { //:")) {
      push(s.slice(12) + " : ({");
    } else {
      li.push(line);
    }
  }
  return li.join("\n");
};
var coffee_plus = (CoffeeScript) => {
  var compile;
  ({
    compile
  } = CoffeeScript);
  compile.bind(CoffeeScript);
  return (code, ...args) => {
    var r;
    code = coffee_label((0, _import2.default)(code.split("\n")));
    r = compile(code, ...args);
    if (typeof r === "string") {
      return label(r);
    } else {
      r.js = label(r.js);
      return r;
    }
  };
};
exports.coffee_plus = coffee_plus;
var _default = (CoffeeScript) => {
  return CoffeeScript.compile = coffee_plus(CoffeeScript);
};
exports.default = _default;
/*! Bundled license information:

lodash-es/lodash.js:
  (**
   * @license
   * Lodash (Custom Build) <https://lodash.com/>
   * Build: `lodash modularize exports="es" -o ./`
   * Copyright OpenJS Foundation and other contributors <https://openjsf.org/>
   * Released under MIT license <https://lodash.com/license>
   * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
   * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
   *)
*/
